uefi/src/proto/network/tcp4.rs

```diff
@@ -0,0 +1,639 @@
+use core::ffi::c_void;
+use core::fmt::Debug;
+use core::ptr::{self, NonNull};
+use core::time::Duration;
+use core::{array, hint};
+use uefi_raw::Boolean;
+use uefi_raw::protocol::driver::ServiceBindingProtocol;
+use uefi_raw::protocol::network::tcp4::{
+    Ip4ModeData, Tcp4AccessPoint, Tcp4CompletionToken, Tcp4ConfigData, Tcp4ConnectionState,
+    Tcp4FragmentData, Tcp4IoToken, Tcp4Option, Tcp4Packet, Tcp4Protocol, Tcp4ReceiveData,
+    Tcp4TransmitData,
+};
+pub use wrappers::{AccessPoint, ConfigData, ConfigOptions};
+
+/// A TCPv4 connection.
+///
+/// # Examples
+///
+/// ```no_run
+/// # fn hello_world() -> uefi::Result {
+/// # extern crate alloc;
+/// use alloc::string::String;
+/// use core::net::Ipv4Addr;
+/// use uefi::{
+///     boot, print, println,
+///     proto::network::tcpv4::{AccessPoint, ConfigData, Tcp4,
+///     Tcp4ServiceBinding,
+/// },
+/// };
+/// use uefi_raw::{
+///     Boolean, protocol::network::tcp4::Tcp4AccessPoint,
+///     protocol::network::tcp4::Tcp4ClientConnectionModeParams,
+/// };
+///
+/// let remote_address = Ipv4Addr::new(192, 0, 2, 2);
+/// let remote_port = 5050;
+///
+/// println!("Connecting to {remote_address:?}:{remote_port}...");
+/// let mut tcp = {
+///     let tcp_svc_handle = boot::get_handle_for_protocol::<Tcp4ServiceBinding>()?;
+///     let mut tcp_svc_proto =
+///         boot::open_protocol_exclusive::<Tcp4ServiceBinding>(tcp_svc_handle)?;
+///     let tcp_proto_handle = tcp_svc_proto.create_child()?;
+///     let mut tcp_proto = boot::open_protocol_exclusive::<Tcp4>(tcp_proto_handle)?;
+///     let config_data = ConfigData {
+///         type_of_service: 0,
+///         time_to_live: 255,
+///         access_point: AccessPoint {
+///             use_default_address: true,
+///             // The following two fields are meaningless when
+///             // `use_default_address == true`
+///             station_address: Ipv4Addr::UNSPECIFIED,
+///             subnet_mask: Ipv4Addr::UNSPECIFIED,
+///             station_port: 0,
+///             remote_address,
+///             remote_port,
+///             // true => client mode
+///             active_flag: true,
+///         },
+///     };
+///     tcp_proto
+///         .configure(&config_data, None)
+///         .expect("configure failed");
+///     tcp_proto.connect()?;
+///     tcp_proto
+/// };
+///
+/// let tx_msg = "Hello";
+/// println!("Sending {tx_msg:?} over TCP...");
+/// tcp.transmit(tx_msg.as_bytes())?;
+///
+/// print!("Received ");
+/// let mut buf = [0_u8; 64];
+/// let n = tcp.receive(&mut buf)?;
+/// let rx_string = String::from_utf8_lossy(&buf[..n]);
+/// println!("{rx_string:?}");
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+#[repr(transparent)]
+#[unsafe_protocol(Tcp4Protocol::GUID)]
+pub struct Tcp4(pub Tcp4Protocol);
+
+impl Tcp4 {
+    /// See [Tcp4Protocol::configure].
+    pub fn configure(&mut self, config: &ConfigData, options: Option<&ConfigOptions>) -> Result {
```

Comment:
I just realize this fn duplicates a lot of `Ip4Config2::ifup()` . Should I just make `Tcp4`  take a NIC Handle that's already been configured like `Http` does?